### ISSUE TITLE
NQR: Fundamental ML Pipeline (FNN / RF / ANFIS) for Next-Quarter Return Prediction

### ISSUE THREAD

**Issue by sudharshan02086** [Rivier-Computer-Science/AI-Agent-Stock-Prediction#597]
**Description:**
Build a CPU-only pipeline that:

- Fetches quarterly financial statements and computes key ratios (ROE, debt/equity, margins, etc.).
- Trains three models—Feed-Forward Neural Network, Random Forest, and ANFIS—on those ratios to predict the sign (up/down) of next-quarter returns.
- Exposes an inference API that emits JSON

```
{ 
  "ticker": "AAPL", 
  "fnn_prob": 0.65, 
  "rf_prob": 0.59, 
  "anfis_prob": 0.71 
}

```
- Uses a Crew AI DecisionAgent to map majority-vote or probability thresholds into a final BUY/SELL recommendation.

**Acceptance Criteria:**

- Ratio Data: Quarterly ratios for each ticker are fetched, cleaned, and stored.
- Model Training:
1. FNN ≥ 60% classification accuracy on validation.
2. RF achieves comparable or better accuracy.
3. ANFIS completes training without errors.
- Inference Output: Given new ratios, the API returns valid JSON with three model probabilities within 1 s.
- Crew AI Decision: A Crew AI agent ingests that JSON and outputs exactly {"recommendation":"BUY"} or {"recommendation":"SELL"}.

**Conditions of Satisfaction:**

- Data Integrity: No missing or malformed ratio inputs; NaNs handled.
- Model Validity: All three models accept the same input shape and produce probabilities in [0, 1].
- Performance: Full inference (load + predict) runs ≤ 1 s on Codespaces CPU.
- Robustness: Pipeline tolerates new tickers or missing quarters without crashing.
- Integration: DecisionAgent always returns BUY or SELL (no HOLD).

**Definition of Done:**

- Ratio fetching and storage scripts implemented and unit-tested.
- FNN, RF, and ANFIS model code written, trained, and validated.
- Inference endpoint loads all three models and emits correct JSON.
- Crew AI decision agent defined and tested end-to-end.
- Integration tests and documentation (README, examples) completed.
- Backtesting harness implemented and performance report generated.

**Agents Overview:**

DataFetchAgent

RatioCalcAgent

ModelTrainAgent

ModelInferAgent

DecisionAgent

**Detailed Agent Descriptions:**

- DataFetchAgent: Fetches quarterly financial statements via API (e.g. Alpha Vantage); emits raw financial fields JSON.
- RatioCalcAgent: Computes ratios (ROE, debt/equity, margins) from raw fields; emits a structured ratios JSON per ticker.
- ModelTrainAgent: Trains FNN, RF, and ANFIS on historical ratio datasets; saves model artifacts and training metrics.
- ModelInferAgent: Loads the three models; given new ratios, outputs { fnn_prob, rf_prob, anfis_prob }.
- DecisionAgent: Reads the three probabilities and returns {"recommendation":"BUY"} if ≥ 2 models ≥ 0.6, else {"recommendation":"SELL"}.

**Tasks & Subtasks:**

- [ ] NQR.1 Ratio Data Retrieval & Preparation
- [ ] NQR.1.1 Implement DataFetchAgent to pull quarterly statements (3 ph)
- [ ] NQR.1.2 Build RatioCalcAgent to compute ratios and handle NaNs (2 ph)
- [ ] NQR.1.3 Write unit tests for data fetching and ratio calculations (1 ph)
- [ ] NQR.2 Model Training Pipeline
- [ ] NQR.2.1 Define FNN architecture and training loop (3 ph)
- [ ] NQR.2.2 Train and validate FNN; log accuracy (2 ph)
- [ ] NQR.2.3 Train Random Forest with hyperparameter tuning (2 ph)
- [ ] NQR.2.4 Train ANFIS model and verify convergence (3 ph)
- [ ] NQR.2.5 Compare validation metrics and select thresholds (1 ph)
- [ ] NQR.3 Inference API Development
- [ ] NQR.3.1 Create ModelInferAgent script to load all three models (2 ph)
- [ ] NQR.3.2 Given input ratios, compute and format { fnn_prob, rf_prob, anfis_prob } JSON (1 ph)
- [ ] NQR.3.3 Benchmark inference latency and optimize if >1 s (1 ph)
- [ ] NQR.4 Crew AI Decision Agent
- [ ] NQR.4.1 Define Crew AI prompt logic for majority-vote or threshold rules (2 ph)
- [ ] NQR.4.2 Implement and test DecisionAgent with edge-case probabilities (1 ph)
- [ ] NQR.5 Integration & Testing
- [ ] NQR.5.1 Write end-to-end integration test: fetch → ratios → infer → decision (2 ph)
- [ ] NQR.5.2 Validate outputs for a sample ticker set (1 ph)
- [ ] NQR.5.3 Document pipeline, usage examples, and agent prompts in README (2 ph)
- [ ] NQR.6 Backtesting & Evaluation
- [ ] NQR.6.1 Prepare historical test sets and data conversion (2 ph)
- [ ] NQR.6.2 Implement backtesting harness to simulate the full infer → decision pipeline over history (3 ph)
- [ ] NQR.6.3 Calculate backtesting performance metrics (accuracy, return, drawdown) and analysis (2 ph)
- [ ] NQR.6.4 Generate backtesting report and visualizations (2 ph)
